src/Simulation/Common/QubitManager.cs

```diff
@@ -74,185 +166,224 @@ public class QubitManager
+            NumAllocatedQubits = 0;
+            NumDisabledQubits = 0;
 
-        private void UpdateQubitCapacity(long newQubitCapacity)
-        {
-            Debug.Assert(newQubitCapacity < MaxQubitCapacity);
-            NumQubits = newQubitCapacity;
-            None = NumQubits;
-            Allocated = NumQubits + 1;
-            Disabled = NumQubits + 2;
-            AllocatedForBorrowing = NumQubits + 3;
-            this.qubits = new long[NumQubits];
-        }
-
-        private void ExtendQubitArray()
-        {
-            long oldNumQubits = NumQubits;
-            long oldNone = None;
-            long oldAllocated = Allocated;
-            long oldDisabled = Disabled;
-            long oldAllocatedForBorrowing = AllocatedForBorrowing;
-            long[] oldQubitsArray = this.qubits;
-
-            UpdateQubitCapacity(oldNumQubits*2); // This changes Allocated, Disabled, AllocatedForBorrowing markers.
-
-            for (long i = 0; i < oldNumQubits; i++)
+            if (!DisableBorrowing)
             {
-                if (oldQubitsArray[i] == oldNone) {
-                    // Point to the first new (free) element
-                    this.qubits[i] = oldNumQubits; 
-                } else if (oldQubitsArray[i] == oldAllocated) {
-                    // Allocated qubits are marked differently now.
-                    this.qubits[i] = Allocated;
-                }
-                else if (oldQubitsArray[i] == oldDisabled)
-                {
-                    // Disabled qubits are marked differently now.
-                    this.qubits[i] = Disabled;
-                }
-                else if (oldQubitsArray[i] >= oldAllocatedForBorrowing) {
-                    // This updates refCounts
-                    this.qubits[i] = (oldQubitsArray[i] - oldAllocatedForBorrowing) + AllocatedForBorrowing ;
-                }
+                operationStack = new Stack<StackFrame>();
+                curFrame = new StackFrame();
             }
+        }
 
-            for (long i = oldNumQubits; i < NumQubits; i++)
-            {
-                this.qubits[i] = i + 1;
-            }
-            Debug.Assert(this.qubits[NumQubits - 1] == None);
+        private class QubitNonAbstract : Qubit
+        {
+            // This class is only needed because Qubit is abstract. 
+            // It is equivalent to Qubit and adds nothing to it except the ability to create it.
+            // It should be used only in CreateQubitObject below, and nowhere else.
+            // When Qubit stops being abstract, this class should be removed.
+            public QubitNonAbstract(int id) : base(id) { }
+        }
 
-            if (free == oldNone)
+        /// <summary>
+        /// May be overriden to create a custom Qubit object of a derived type.
+        /// </summary>
+        /// <param name="id">unique qubit id</param>
+        /// <returns>a newly instantiated qubit</returns>
+        public virtual Qubit CreateQubitObject(long id)
```

Comment:
Then the conversion long to int just needs to happen in a bunch of other places (since Q# has Int64 by default). Hence having it throw here seems somewhat cleaner to me in this case. The alternative would be to make the entire qubit manager int based, and then have the checks at the intersection with Q# (meaning Allocate, Borrow, GetQubitsAvailableToUse, and GetQubitsAvailableToBorrow would need to do this check).

Reply:
OK, makes sense.